### PR TITLE
New api get email receive update

### PR DESCRIPTION
**Pull request by vunquitk11**
- fix conflict with 2 old branch api
- doing api get-common-friends
- doing api get-emails-receive-update

### CODE REVIEW

File: api/internal/controller/relations/list.go [hunk 1]
```diff
@@ -29,3 +29,48 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
 }
+
+// GetCommonFriendList retrieve the common friends list between two email addresses.
+func (i impl) GetCommonFriendList(ctx context.Context, friends model.GetCommonFriendListInput) ([]string, error) {
+
+	firstID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, friends.FirstEmail)
+
+	if err != nil {
+
+		if err == sql.ErrNoRows {
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:40]: Check sql.ErrNoRows at repository

File: api/internal/controller/relations/list.go [hunk 2]
```diff
@@ -29,3 +29,48 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
+		}
+
+		return nil, err
+	}
+
+	secondID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, friends.SecondEmail)
+
+	if err != nil {
+
+		if err == sql.ErrNoRows {
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:51]: Check sql.ErrNoRows at repository

File: api/internal/controller/relations/list.go [hunk 3]
```diff
@@ -29,3 +29,48 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
+}
+
+// GetEmailsReceiveUpdate retrieve the common friends list between two email addresses.
+func (i impl) GetEmailsReceiveUpdate(ctx context.Context, sender string, text string) ([]string, error) {
+
+	userID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, sender)
+
+	if err != nil {
+
+		if err == sql.ErrNoRows {
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:68]: Check sql.ErrNoRows at repository

File: api/internal/handler/gql/mod/utils.go [hunk 4]
```diff
@@ -15,3 +17,46 @@ func IsValidEmail(email string) (bool, error) {
...
+	for _, item := range b {
+		if _, ok := m[item]; ok {
+			c = append(c, item)
+		}
+	}
+	return
+}
+
+// Remove index in slice
+func RemoveIndex(s []int64, index int) []int64 {
```
Comment (vantran95) [inline api/internal/handler/gql/mod/utils.go:38]: This func is an internal func, right?

File: api/internal/handler/gql/public/gqlgenerated.go [hunk 5]
```diff
@@ -0,0 +1,5769 @@
+// Code generated by github.com/99designs/gqlgen, DO NOT EDIT.
```
Comment (vantran95) [inline api/internal/handler/gql/public/gqlgenerated.go:1]: Again, Don't push this into git

File: api/internal/handler/gql/public/models_gen.go [hunk 6]
```diff
@@ -0,0 +1,3 @@
+// Code generated by github.com/99designs/gqlgen, DO NOT EDIT.
+
+package public
```
Comment (vantran95) [inline api/internal/handler/gql/public/models_gen.go:3]: Again, Don't push this into git

File: api/internal/repository/inventory/relations.go [hunk 7]
```diff
@@ -170,6 +241,69 @@ func (i impl) GetFriendList(ctx context.Context, userID int64) ([]string, error)
 	return emails, nil
 }
 
+// GetCommonFriendList retrieve the common friends list between two email addresses from DB
+func (i impl) GetCommonFriendList(ctx context.Context, firstID, secondID int64) ([]string, error) {
+
+	if firstID == secondID {
```
Comment (vantran95) [inline api/internal/repository/inventory/relations.go:247]: Need to check this condition with another API createRealtion

File: api/internal/repository/inventory/relations.go [hunk 8]
```diff
@@ -170,6 +241,69 @@ func (i impl) GetFriendList(ctx context.Context, userID int64) ([]string, error)
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	friendsOfSecondUser, err := GetFriendsIdOfUser(ctx, &i.dbConn, secondID)
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (vantran95) [inline api/internal/repository/inventory/relations.go:261]: I think you should move this code into the controller

File: api/internal/repository/inventory/relations.go [hunk 9]
```diff
@@ -170,6 +241,69 @@ func (i impl) GetFriendList(ctx context.Context, userID int64) ([]string, error)
...
+
+	subsID, err := GetRelationRequesters(ctx, &i.dbConn, userID, model.RelationStatusSub.String())
+
+	if err != nil {
+		return nil, err
+	}
```
Comment (vantran95) [inline api/internal/repository/inventory/relations.go:288]: I think you should move these codes into the controller

File: api/internal/repository/inventory/relations.go [hunk 10]
```diff
@@ -170,6 +241,69 @@ func (i impl) GetFriendList(ctx context.Context, userID int64) ([]string, error)
 	return emails, nil
 }
 
+// GetCommonFriendList retrieve the common friends list between two email addresses from DB
+func (i impl) GetCommonFriendList(ctx context.Context, firstID, secondID int64) ([]string, error) {
```
Comment (vantran95) [inline api/internal/repository/inventory/relations.go:245]: This function doesn't have any direct impact on DB.
Should move to the controller.

add repository only have function have a query with the database

File: api/internal/repository/inventory/relations.go [hunk 11]
```diff
@@ -170,6 +241,69 @@ func (i impl) GetFriendList(ctx context.Context, userID int64) ([]string, error)
...
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	return emails, nil
+}
+
+// GetEmailsReceiveUpdate retrieve emails can receive update from an email
+func (i impl) GetEmailsReceiveUpdate(ctx context.Context, userID int64, text string) ([]string, error) {
```
Comment (vantran95) [inline api/internal/repository/inventory/relations.go:276]: This function doesn't have any direct impact on DB.
Should move to the controller.

add repository only have function have a query with the database

File: api/internal/repository/inventory/relations.go [hunk 12]
```diff
@@ -252,7 +386,7 @@ func (i impl) CreateBlockRelation(ctx context.Context, requesterID int64, target
 	}
 
 	if err := o.Insert(ctx, &i.dbConn, boil.Infer()); err != nil {
-		return false, pkgerrors.WithStack(err)
+		return false, err
```
Comment (vantran95) [inline api/internal/repository/inventory/relations.go:389]:
```suggestion
		return false, pkgerrors.WithStack(err)
```

File: .gitignore [hunk 13]
```diff
@@ -2,5 +2,6 @@
 /shelf/
 # Editor-based HTTP Client requests
 /httpRequests/
+api/internal/handler/gql/public/models_gen.go
 api/internal/handler/gql/public/gqlgenerated.go
-.idea
+.idea
```
Comment (vantran95) [inline .gitignore:7]: Add new space line

File: api/internal/controller/relations/list.go [hunk 14]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
 	userID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, email)
 
 	if err != nil {
+		return nil, err
+	}
 
-		if err == sql.ErrNoRows {
-			return nil, inventory.ErrUserNotFound
-		}
+	friendsID, err := i.repo.Inventory().GetFriendsIdOfUser(ctx, userID)
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:20]:
```suggestion
	friendIDs, err := i.repo.Inventory().GetFriendsIdOfUser(ctx, userID)
```

File: api/internal/controller/relations/list.go [hunk 15]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
 	userID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, email)
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:14]: Check case userID = 0

File: api/internal/controller/relations/list.go (hunk 14 above)
Comment (vantran95) [inline api/internal/controller/relations/list.go:20]: Check case friendIDs == 0

File: api/internal/controller/relations/list.go [hunk 16]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
-			return nil, inventory.ErrUserNotFound
-		}
+	friendsID, err := i.repo.Inventory().GetFriendsIdOfUser(ctx, userID)
 
+	if err != nil {
 		return nil, err
 	}
 
-	return i.repo.Inventory().GetFriendList(ctx, userID)
+	emails, err := i.repo.Inventory().GetEmailByIDs(ctx, friendsID)
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:26]: Check case len(emails) == 0

File: api/internal/controller/relations/list.go [hunk 17]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
+		return nil, err
+	}
+
+	friendsOfSecondUser, err := i.repo.Inventory().GetFriendsIdOfUser(ctx, secondID)
+
+	if err != nil {
+		return nil, err
+	}
+
+	emails, err := i.repo.Inventory().GetEmailByIDs(ctx, mod.Intersection(friendsOfFirstUser, friendsOfSecondUser))
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:71]: Check case len(emails) == 0

File: api/internal/controller/relations/list.go [hunk 18]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
 
 // List get all relations record
 func (i impl) GetRelationList(ctx context.Context) ([]model.Relation, error) {
 	return i.repo.Inventory().GetRelationList(ctx)
 }
+
+// GetCommonFriendList retrieve the common friends list between two email addresses.
+func (i impl) GetCommonFriendList(ctx context.Context, friends model.GetCommonFriendListInput) ([]string, error) {
+
+	firstID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, friends.FirstEmail)
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:43]: Check case firstID == 0

File: api/internal/controller/relations/list.go [hunk 19]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	secondID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, friends.SecondEmail)
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:49]: Check case secondID == 0

File: api/internal/controller/relations/list.go [hunk 20]
```diff
@@ -14,18 +14,111 @@ func (i impl) GetFriendList(ctx context.Context, email string) ([]string, error)
...
+		return nil, err
+	}
+
+	return emails, nil
+}
+
+// GetEmailsReceiveUpdate retrieve the common friends list between two email addresses.
+func (i impl) GetEmailsReceiveUpdate(ctx context.Context, sender string, text string) ([]string, error) {
+
+	userID, err := i.repo.Inventory().GetUserIDFromEmail(ctx, sender)
```
Comment (vantran95) [inline api/internal/controller/relations/list.go:83]: Please check case ID == 0 when use func `GetUserIDFromEmail` & `GetFriendsIdOfUser`

File: api/internal/handler/gql/public/gqlgenerated.go (hunk 5 above)
Comment (vantran95) [inline api/internal/handler/gql/public/gqlgenerated.go:1]: Again, Please add this file into .gitignore

File: api/internal/repository/inventory/users.go [hunk 21]
```diff
@@ -12,44 +12,74 @@ import (
...
+		qm.Select(orm.UserColumns.ID, orm.UserColumns.Email),
+		qm.WhereIn(orm.UserColumns.ID+" in ?", args...),
+	).All(ctx, &i.dbConn)
 
 	if err != nil {
-		if err == sql.ErrNoRows {
-			return 0, ErrUserNotFound
-		}
-		return 0, err
+		return nil, pkgerrors.WithStack(err)
```
Comment (vantran95) [inline api/internal/repository/inventory/users.go:29]: Check case sql.ErrNoRows

File: api/internal/repository/inventory/users.go [hunk 22]
```diff
@@ -12,44 +12,74 @@ import (
...
 
 	slice, err := orm.Users(
-		qm.WhereIn(orm.UserColumns.ID+" in ?", args...),
-	).All(ctx, dbConn)
+		qm.Select(orm.UserColumns.ID, orm.UserColumns.Email),
+		qm.WhereIn(orm.UserColumns.Email+" in ?", args...),
+	).All(ctx, &i.dbConn)
 
 	if err != nil {
 		return nil, pkgerrors.WithStack(err)
```
Comment (vantran95) [inline api/internal/repository/inventory/users.go:54]: Check case sql.ErrNoRows

Comment (vunquitk11) [reply api/internal/repository/inventory/users.go:54]: The api using this function has the case of returning an empty array, if it returns errNotFound, then when mapping the data to the response model, I have to compare the error if it is errNotFound, then assign the empty array to return. So I will not edit this issue.

File: .gitignore [hunk 23]
```diff
@@ -2,5 +2,6 @@
 /shelf/
 # Editor-based HTTP Client requests
 /httpRequests/
+api/internal/handler/gql/public/models_gen.go
```
Comment (vantran95) [inline .gitignore:5]: Don't add this file into .gitignore 

This file should not be generated by the gqlgen,

File: api/internal/controller/relations/create.go [hunk 24]
```diff
@@ -6,8 +6,29 @@ import (
 	"github.com/S3_FriendManagement_VuNguyen/api/internal/model"
 )
 
+<<<<<<< HEAD
```
Comment (vantran95) [inline api/internal/controller/relations/create.go:9]: Resolve conflic

**Pull request merged**